src/util/psbt/roles.rs

```diff
@@ -0,0 +1,373 @@
+    fn add_output_script(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error>;
+
+    /// Adds P2WSH script source for a given output index
+    fn add_output_wscript(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error>;
+
+    /// Adds information about public key derivation for a given output index
+    /// (for P2PK and P2WPK outputs)
+    fn add_output_derivation(&mut self, index: u32, pubkey: &PublicKey, fingerprint: Fingerprint, path: &DerivationPath) -> Result<&mut Self, Error>;
+}
+
+impl Updater for PartiallySignedTransaction {
+    fn has_sign_info(&self) -> bool {
+        unimplemented!()
+    }
+
+    fn add_input_tx(&mut self, index: u32, tx: &Transaction) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_input_utxo(&mut self, index: u32, txin: &TxIn) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_input_script(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_input_wscript(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_input_derivation(&mut self, index: u32, pubkey: &PublicKey, fingerprint: Fingerprint, path: &DerivationPath) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_output_script(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_output_wscript(&mut self, index: u32, script: &Script) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+
+    fn add_output_derivation(&mut self, index: u32, pubkey: &PublicKey, fingerprint: Fingerprint, path: &DerivationPath) -> Result<&mut Self, Error> {
+        unimplemented!()
+    }
+}
+
+/// Errors which may happen during signer verification
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum SignerVerificationError {
+}
+
+impl fmt::Display for SignerVerificationError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        (self as &fmt::Debug).fmt(f)
+    }
+}
+
+/// Implementation of [Role::Signer] role according to
+/// https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#Signer
+pub trait Signer {
+    /// Detects when at least some of the signatures are present
+    fn has_partial_signatures(&self) -> bool;
+
+    /// Verifies that PSBT is ready to be signed, i.e. contains all necessary
+    /// information
+    fn verify(&self) -> Vec<SignerVerificationError>;
+
+    /// Adds signature for a specific input
+    fn add_signature(&mut self, input: u32, signature: Signature) -> Result<&mut Self, Error>;
+}
+
+impl Signer for PartiallySignedTransaction {
+    fn has_partial_signatures(&self) -> bool {
+        unimplemented!()
+    }
+
+    fn verify(&self) -> Vec<SignerVerificationError> {
+        unimplemented!()
+    }
+
+    fn add_signature(&mut self, input: u32, signature: Signature) -> Result<&mut Self, Error> {
```

Comment:
Most probably if `trait Signer` is useful, it should be something like:

```
fn sign(tx: PSBT) -> PSBT;
```

The signer is responsible for coming up with the signature, and can't expect to be given it. A `struct MyKeyIsOnTheDisk` will produce signature from a secret key, and `HSMSigner` will ask HSM to do it or something. Any neccessary arguments would be provided in a constructor of the concrete type implementing `trait Signer` .

Reply:
Pls see my comment above; this is not a generic Signer but a Signer role cross-section of PSBT (used as `roles::Signer`). There will be a `Sync/AsyncSigner` trait outside of the role scope that will be taking PSBT in a `Signer` role as an arbument.

What we can think is on renaming roles into adjectives, like `Signer -> Signable` and `Combiner -> Combinable`

Reply:
Oh. These names would be much better.